File: backend/src/firefighter_tools_backend/routes/calendar_converter.py
"""HTTP endpoints for converting schedules and downloading the example."""

from pathlib import Path, PurePath

from fastapi import APIRouter, Depends, File, Query, UploadFile, status
from fastapi.responses import FileResponse, JSONResponse
from sqlalchemy.orm import Session

from firefighter_tools_backend.dependencies import (
    get_current_user,
    get_db,
    require_super_user,
)
from firefighter_tools_backend.domain.calendar_conversion import (
    ConversionError,
    ConversionErrorCode,
    ConversionResult,
)
from firefighter_tools_backend.domain.schedule_store import (
    ScheduleStoreError,
    ScheduleStoreErrorCode,
    StoredSchedule,
)
from firefighter_tools_backend.domain.upload import (
    UploadValidationError,
    UploadValidationErrorCode,
)
from firefighter_tools_backend.domain.user import (
    AuthError,
    AuthErrorCode,
    Role,
    User,
)
from firefighter_tools_backend.models.auth import AuthErrorResponse
from firefighter_tools_backend.models.calendar_conversion import (
    ActiveSchedule,
    ActiveScheduleResponse,
    Calendar,
    ConversionResponse,
    ConversionScope,
    ConversionStatus,
    FatalErrorCode,
    FatalErrorResponse,
    InvalidEvent,
    SourcePosition,
)
from firefighter_tools_backend.services import calendar_conversion, schedule_store
from firefighter_tools_backend.services.upload_validation import validate_upload

router = APIRouter(prefix="/tools/calendar-converter", tags=["calendar converter"])

_SAMPLE_SCHEDULE_PATH = (
    Path(__file__).parents[4]
    / "assets"
    / "examples"
    / "calendar_schedule_example.xlsx"
)
_SAMPLE_SCHEDULE_MEDIA_TYPE = (
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
)

_UPLOAD_ERRORS = {
    UploadValidationErrorCode.MISSING_FILENAME: (
        status.HTTP_422_UNPROCESSABLE_CONTENT,
        FatalErrorCode.MISSING_FILENAME,
        "A filename is required.",
    ),
    UploadValidationErrorCode.UNSUPPORTED_FILE_TYPE: (
        status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
        FatalErrorCode.UNSUPPORTED_FILE_TYPE,
        "Only CSV and XLSX files are supported.",
    ),
    UploadValidationErrorCode.OVERSIZED_UPLOAD: (
        status.HTTP_413_CONTENT_TOO_LARGE,
        FatalErrorCode.OVERSIZED_UPLOAD,
        "The uploaded file exceeds the 10 MiB limit.",
    ),
    UploadValidationErrorCode.INPUT_READ_ERROR: (
        status.HTTP_422_UNPROCESSABLE_CONTENT,
        FatalErrorCode.INPUT_READ_ERROR,
        "The uploaded file could not be read.",
    ),
}

_CONVERSION_ERRORS = {
    ConversionErrorCode.UNSUPPORTED_FILE_TYPE: (
        status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
        FatalErrorCode.UNSUPPORTED_FILE_TYPE,
        "Only CSV and XLSX files are supported.",
    ),
    ConversionErrorCode.MALFORMED_CSV: (
        status.HTTP_422_UNPROCESSABLE_CONTENT,
        FatalErrorCode.MALFORMED_CSV,
        "The CSV schedule is malformed.",
    ),
    ConversionErrorCode.MALFORMED_XLSX: (
        status.HTTP_422_UNPROCESSABLE_CONTENT,
        FatalErrorCode.MALFORMED_XLSX,
        "The XLSX schedule is malformed.",
    ),
    ConversionErrorCode.INPUT_READ_ERROR: (
        status.HTTP_422_UNPROCESSABLE_CONTENT,
        FatalErrorCode.INPUT_READ_ERROR,
        "The uploaded file could not be read.",
    ),
}

_STORE_ERRORS = {
    ScheduleStoreErrorCode.NO_ACTIVE_SCHEDULE: (
        status.HTTP_409_CONFLICT,
        FatalErrorCode.NO_ACTIVE_SCHEDULE,
        "No schedule has been uploaded yet.",
    ),
    ScheduleStoreErrorCode.STORE_WRITE_ERROR: (
        status.HTTP_500_INTERNAL_SERVER_ERROR,
        FatalErrorCode.INTERNAL_ERROR,
        "The request could not be processed.",
    ),
}

_MISSING_PERSONNEL_NUMBER = (
    status.HTTP_409_CONFLICT,
    FatalErrorCode.MISSING_PERSONNEL_NUMBER,
    "This account has no personnel number.",
)


@router.get(
    "/example",
    response_class=FileResponse,
    responses={401: {"model": AuthErrorResponse}},
)
def download_example_schedule(
    _: User = Depends(get_current_user),
) -> FileResponse:
    """Download the version-controlled example schedule without copying it."""
    return FileResponse(
        _SAMPLE_SCHEDULE_PATH,
        media_type=_SAMPLE_SCHEDULE_MEDIA_TYPE,
        filename=_SAMPLE_SCHEDULE_PATH.name,
    )


@router.put(
    "/schedule",
    response_model=ActiveScheduleResponse,
    responses={
        401: {"model": AuthErrorResponse},
        403: {"model": AuthErrorResponse},
        413: {"model": FatalErrorResponse},
        415: {"model": FatalErrorResponse},
        422: {"model": FatalErrorResponse},
        500: {"model": FatalErrorResponse},
    },
)
async def replace_active_schedule(
    file: UploadFile | None = File(default=None),
    current_user: User = Depends(require_super_user),
    session: Session = Depends(get_db),
) -> ActiveScheduleResponse | JSONResponse:
    """Store one validated upload as the schedule every account converts."""
    if file is None:
        return _fatal_response(
            *_UPLOAD_ERRORS[UploadValidationErrorCode.MISSING_FILENAME]
        )

    try:
        upload = await validate_upload(file)
        schedule = schedule_store.save_active_schedule(
            session,
            upload,
            uploaded_by=current_user,
        )
        return _active_schedule_response(schedule)
    except UploadValidationError as error:
        return _fatal_response(*_UPLOAD_ERRORS[error.code])
    except ScheduleStoreError as error:
        return _fatal_response(*_STORE_ERRORS[error.code])
    except Exception:
        return _fatal_response(
            status.HTTP_500_INTERNAL_SERVER_ERROR,
            FatalErrorCode.INTERNAL_ERROR,
            "The request could not be processed.",
        )


@router.get(
    "/schedule",
    response_model=ActiveScheduleResponse,
    responses={401: {"model": AuthErrorResponse}},
)
def read_active_schedule(
    _: User = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> ActiveScheduleResponse:
    """Report the stored schedule, or ``None`` when none is held yet."""
    schedule = schedule_store.find_active_schedule(session)
    if schedule is None:
        return ActiveScheduleResponse(schedule=None)
    return _active_schedule_response(schedule)


@router.post(
    "/schedule/convert",
    response_model=ConversionResponse,
    responses={
        401: {"model": AuthErrorResponse},
        403: {"model": AuthErrorResponse},
        409: {"model": FatalErrorResponse},
        415: {"model": FatalErrorResponse},
        422: {"model": FatalErrorResponse},
        500: {"model": FatalErrorResponse},
    },
)
def convert_active_schedule(
    scope: ConversionScope = Query(
        ConversionScope.PERSONAL,
        description=(
            "`personal` returns the caller's events plus the events for "
            "everyone; `full` returns the whole schedule (super-user only)."
        ),
    ),
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> ConversionResponse | JSONResponse:
    """Convert the stored schedule for the signed-in account.

    A personal conversion matches the account's personnel number against the
    schedule's ``participants`` column, so an account without a number is
    refused rather than silently given only the events for everyone.
    """
    if scope is ConversionScope.FULL:
        if current_user.role is not Role.SUPER_USER:
            raise AuthError(AuthErrorCode.FORBIDDEN)
        participant = None
    else:
        participant = current_user.personnel_number
        if participant is None:
            return _fatal_response(*_MISSING_PERSONNEL_NUMBER)

    try:
        schedule = schedule_store.load_active_schedule(session)
        with schedule_store.open_active_schedule(schedule) as source:
            result = calendar_conversion.convert_calendar(
                source,
                filename=schedule.original_filename,
                participant=participant,
            )
        return _conversion_response(
            result,
            source_filename=schedule.original_filename,
            filename_suffix=participant,
        )
    except ScheduleStoreError as error:
        return _fatal_response(*_STORE_ERRORS[error.code])
    except ConversionError as error:
        return _fatal_response(*_CONVERSION_ERRORS[error.code])
    except Exception:
        return _fatal_response(
            status.HTTP_500_INTERNAL_SERVER_ERROR,
            FatalErrorCode.INTERNAL_ERROR,
            "The request could not be processed.",
        )


@router.post(
    "/convert",
    response_model=ConversionResponse,
    responses={
        401: {"model": AuthErrorResponse},
        403: {"model": AuthErrorResponse},
        413: {"model": FatalErrorResponse},
        415: {"model": FatalErrorResponse},
        422: {"model": FatalErrorResponse},
        500: {"model": FatalErrorResponse},
    },
)
async def convert_calendar_upload(
    file: UploadFile | None = File(default=None),
    _: User = Depends(require_super_user),
) -> ConversionResponse | JSONResponse:
    """Validate one upload, convert it, and expose only declared responses."""
    if file is None:
        return _fatal_response(
            *_UPLOAD_ERRORS[UploadValidationErrorCode.MISSING_FILENAME]
        )

    try:
        upload = await validate_upload(file)
        result = calendar_conversion.convert_calendar(
            upload.source,
            filename=upload.filename,
        )
        return _conversion_response(result, source_filename=upload.filename)
    except UploadValidationError as error:
        return _fatal_response(*_UPLOAD_ERRORS[error.code])
    except ConversionError as error:
        return _fatal_response(*_CONVERSION_ERRORS[error.code])
    except Exception:
        return _fatal_response(
            status.HTTP_500_INTERNAL_SERVER_ERROR,
            FatalErrorCode.INTERNAL_ERROR,
            "The request could not be processed.",
        )


def _conversion_response(
    result: ConversionResult,
    *,
    source_filename: str,
    filename_suffix: str | None = None,
) -> ConversionResponse:
    if result.converted_count == 0:
        conversion_status = ConversionStatus.FAILURE
        calendar = None
    else:
        conversion_status = (
            ConversionStatus.PARTIAL
            if result.skipped_count
            else ConversionStatus.SUCCESS
        )
        # A personnel number is restricted to filename-safe characters.
        stem = PurePath(source_filename).stem
        if filename_suffix is not None:
            stem = f"{stem}-{filename_suffix}"
        calendar = Calendar(
            filename=f"{stem}.ics",
            ics_text=result.ics_text,
        )

    return ConversionResponse(
        status=conversion_status,
        total_count=result.total_count,
        converted_count=result.converted_count,
        skipped_count=result.skipped_count,
        invalid_events=[
            InvalidEvent(
                source_position=SourcePosition(
                    event_index=event.source_position.event_index,
                    row=event.source_position.row,
                    worksheet=event.source_position.worksheet,
                ),
                id=event.id,
                summary=event.summary,
                issue_codes=list(event.issue_codes),
            )
            for event in result.invalid_events
        ],
        calendar=calendar,
    )


def _fatal_response(
    status_code: int,
    code: FatalErrorCode,
    message: str,
) -> JSONResponse:
    payload = FatalErrorResponse(code=code, message=message)
    return JSONResponse(
        status_code=status_code,
        content=payload.model_dump(mode="json"),
    )


def _active_schedule_response(schedule: StoredSchedule) -> ActiveScheduleResponse:
    """Expose the stored schedule without its opaque on-disk filename."""
    return ActiveScheduleResponse(
        schedule=ActiveSchedule(
            filename=schedule.original_filename,
            size_bytes=schedule.size_bytes,
            uploaded_at=schedule.uploaded_at,
            uploaded_by=schedule.uploaded_by,
        )
    )

File: frontend/src/api/calendarConverter.ts
export const CALENDAR_CONVERTER_ENDPOINT =
  "/api/v1/tools/calendar-converter/convert";
export const ACTIVE_SCHEDULE_ENDPOINT =
  "/api/v1/tools/calendar-converter/schedule";
export const ACTIVE_SCHEDULE_CONVERT_ENDPOINT =
  "/api/v1/tools/calendar-converter/schedule/convert";

export type ActiveSchedule = {
  filename: string;
  size_bytes: number;
  uploaded_at: string;
  uploaded_by: string;
};

export type ActiveScheduleResponse = {
  schedule: ActiveSchedule | null;
};

export type ActiveScheduleResult =
  | { ok: true; response: ActiveScheduleResponse }
  | { ok: false; status: number; error: FatalErrorResponse };

export type ConversionStatus = "success" | "partial" | "failure";

export type ApiErrorCode =
  | "missing_filename"
  | "unsupported_file_type"
  | "oversized_upload"
  | "malformed_csv"
  | "malformed_xlsx"
  | "input_read_error"
  | "no_active_schedule"
  | "missing_personnel_number"
  | "internal_error";

/**
 * Which events a conversion of the active schedule returns: `personal` keeps
 * the signed-in account's events plus the events for everyone, `full` returns
 * the whole schedule and is reserved for a super-user.
 */
export type ConversionScope = "personal" | "full";

export type ConverterIssueCode =
  | "empty_id"
  | "empty_summary"
  | "end_date_before_start_date"
  | "end_time_before_start_time"
  | "duplicate_id";

export type SourcePosition = {
  event_index: number;
  row: number;
  worksheet: string | null;
};

export type InvalidEvent = {
  source_position: SourcePosition;
  id: string;
  summary: string;
  issue_codes: ConverterIssueCode[];
};

export type Calendar = {
  filename: string;
  mime_type: "text/calendar;charset=utf-8";
  ics_text: string;
};

export type ConversionResponse = {
  status: ConversionStatus;
  total_count: number;
  converted_count: number;
  skipped_count: number;
  invalid_events: InvalidEvent[];
  calendar: Calendar | null;
};

export type FatalErrorResponse = {
  code: ApiErrorCode;
  message: string;
};

export type CalendarConverterResult =
  | {
      ok: true;
      response: ConversionResponse;
    }
  | {
      ok: false;
      status: number;
      error: FatalErrorResponse;
    };

export class CalendarConverterContractError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CalendarConverterContractError";
  }
}

type ConvertCalendarOptions = {
  signal?: AbortSignal;
};

const conversionStatuses = new Set<ConversionStatus>([
  "success",
  "partial",
  "failure",
]);

const apiErrorCodes = new Set<ApiErrorCode>([
  "missing_filename",
  "unsupported_file_type",
  "oversized_upload",
  "malformed_csv",
  "malformed_xlsx",
  "input_read_error",
  "no_active_schedule",
  "missing_personnel_number",
  "internal_error",
]);

const converterIssueCodes = new Set<ConverterIssueCode>([
  "empty_id",
  "empty_summary",
  "end_date_before_start_date",
  "end_time_before_start_time",
  "duplicate_id",
]);

export async function convertCalendar(
  file: File,
  options: ConvertCalendarOptions = {},
): Promise<CalendarConverterResult> {
  const formData = new FormData();
  formData.append("file", file);

  const httpResponse = await fetch(CALENDAR_CONVERTER_ENDPOINT, {
    method: "POST",
    body: formData,
    signal: options.signal,
  });
  const payload: unknown = await readJson(httpResponse);

  if (httpResponse.ok) {
    if (!isConversionResponse(payload)) {
      throw new CalendarConverterContractError(
        "The calendar converter returned an invalid success response.",
      );
    }

    return {
      ok: true,
      response: payload,
    };
  }

  if (!isFatalErrorResponse(payload)) {
    throw new CalendarConverterContractError(
      "The calendar converter returned an invalid error response.",
    );
  }

  return {
    ok: false,
    status: httpResponse.status,
    error: payload,
  };
}

async function readJson(response: Response): Promise<unknown> {
  try {
    return await response.json();
  } catch {
    throw new CalendarConverterContractError(
      "The calendar converter returned a non-JSON response.",
    );
  }
}

function isConversionResponse(value: unknown): value is ConversionResponse {
  if (!isRecord(value)) {
    return false;
  }

  const status = value.status;
  const totalCount = value.total_count;
  const convertedCount = value.converted_count;
  const skippedCount = value.skipped_count;
  const invalidEvents = value.invalid_events;
  const calendar = value.calendar;

  if (
    typeof status !== "string" ||
    !conversionStatuses.has(status as ConversionStatus) ||
    !isNonNegativeInteger(totalCount) ||
    !isNonNegativeInteger(convertedCount) ||
    !isNonNegativeInteger(skippedCount) ||
    !Array.isArray(invalidEvents) ||
    !invalidEvents.every(isInvalidEvent) ||
    !(calendar === null || isCalendar(calendar)) ||
    totalCount !== convertedCount + skippedCount ||
    skippedCount !== invalidEvents.length
  ) {
    return false;
  }

  if (status === "success") {
    return skippedCount === 0 && calendar !== null;
  }
  if (status === "partial") {
    return convertedCount > 0 && skippedCount > 0 && calendar !== null;
  }
  return convertedCount === 0 && calendar === null;
}

function isInvalidEvent(value: unknown): value is InvalidEvent {
  if (!isRecord(value)) {
    return false;
  }

  return (
    isSourcePosition(value.source_position) &&
    typeof value.id === "string" &&
    typeof value.summary === "string" &&
    Array.isArray(value.issue_codes) &&
    value.issue_codes.length > 0 &&
    value.issue_codes.every(
      (code) =>
        typeof code === "string" &&
        converterIssueCodes.has(code as ConverterIssueCode),
    )
  );
}

function isSourcePosition(value: unknown): value is SourcePosition {
  if (!isRecord(value)) {
    return false;
  }

  return (
    isPositiveInteger(value.event_index) &&
    isPositiveInteger(value.row) &&
    (value.worksheet === null || typeof value.worksheet === "string")
  );
}

function isCalendar(value: unknown): value is Calendar {
  if (!isRecord(value)) {
    return false;
  }

  return (
    typeof value.filename === "string" &&
    value.filename.length > 0 &&
    value.mime_type === "text/calendar;charset=utf-8" &&
    typeof value.ics_text === "string"
  );
}

function isFatalErrorResponse(value: unknown): value is FatalErrorResponse {
  if (!isRecord(value)) {
    return false;
  }

  return (
    typeof value.code === "string" &&
    apiErrorCodes.has(value.code as ApiErrorCode) &&
    typeof value.message === "string" &&
    value.message.length > 0
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isNonNegativeInteger(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value) && value >= 0;
}

function isPositiveInteger(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value) && value >= 1;
}

export async function fetchActiveSchedule(
  options: ConvertCalendarOptions = {},
): Promise<ActiveScheduleResult> {
  const httpResponse = await fetch(ACTIVE_SCHEDULE_ENDPOINT, {
    signal: options.signal,
  });
  return activeScheduleResult(httpResponse);
}

export async function uploadActiveSchedule(
  file: File,
  options: ConvertCalendarOptions = {},
): Promise<ActiveScheduleResult> {
  const formData = new FormData();
  formData.append("file", file);

  const httpResponse = await fetch(ACTIVE_SCHEDULE_ENDPOINT, {
    method: "PUT",
    body: formData,
    signal: options.signal,
  });
  return activeScheduleResult(httpResponse);
}

export async function convertActiveSchedule(
  options: ConvertCalendarOptions & { scope?: ConversionScope } = {},
): Promise<CalendarConverterResult> {
  const scope = options.scope ?? "personal";
  const endpoint = `${ACTIVE_SCHEDULE_CONVERT_ENDPOINT}?scope=${scope}`;
  const httpResponse = await fetch(endpoint, {
    method: "POST",
    signal: options.signal,
  });
  const payload: unknown = await readJson(httpResponse);

  if (httpResponse.ok) {
    if (!isConversionResponse(payload)) {
      throw new CalendarConverterContractError(
        "The calendar converter returned an invalid success response.",
      );
    }

    return { ok: true, response: payload };
  }

  if (!isFatalErrorResponse(payload)) {
    throw new CalendarConverterContractError(
      "The calendar converter returned an invalid error response.",
    );
  }

  return { ok: false, status: httpResponse.status, error: payload };
}

async function activeScheduleResult(
  httpResponse: Response,
): Promise<ActiveScheduleResult> {
  const payload: unknown = await readJson(httpResponse);

  if (httpResponse.ok) {
    if (!isActiveScheduleResponse(payload)) {
      throw new CalendarConverterContractError(
        "The calendar converter returned an invalid schedule response.",
      );
    }

    return { ok: true, response: payload };
  }

  if (!isFatalErrorResponse(payload)) {
    throw new CalendarConverterContractError(
      "The calendar converter returned an invalid error response.",
    );
  }

  return { ok: false, status: httpResponse.status, error: payload };
}

function isActiveScheduleResponse(
  value: unknown,
): value is ActiveScheduleResponse {
  if (!isRecord(value)) {
    return false;
  }

  const { schedule } = value;
  return schedule === null || isActiveSchedule(schedule);
}

function isActiveSchedule(value: unknown): value is ActiveSchedule {
  if (!isRecord(value)) {
    return false;
  }

  return (
    typeof value.filename === "string" &&
    value.filename.length > 0 &&
    isNonNegativeInteger(value.size_bytes) &&
    typeof value.uploaded_by === "string" &&
    value.uploaded_by.length > 0 &&
    typeof value.uploaded_at === "string" &&
    Number.isFinite(Date.parse(value.uploaded_at))
  );
}
